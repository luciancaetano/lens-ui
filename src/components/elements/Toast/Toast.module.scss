@import '../../../styles/lib';
@import '../../../styles/typography';

.clickable {
  pointer-events: all;

  * {
    pointer-events: all;
  }
}

.toast {
  @include lens-ui-typography;

  pointer-events: all;
  min-width: 300px;
  border-radius: .1rem;
  margin-top: .3rem;
  box-shadow: var(--lens-ui-elements-card-box-shadow);
  display: flex;
  flex-direction: column;
  overflow: hidden;
  z-index: map-get($layers, 'Toast');

  &--mobile {
    width: 100%;
  }

  [data-lens-element='progress-bar'],
  [data-lens-element='progress-bar__indicator'] {
    border-radius: 0;
    height: .3rem;
  }

  &__main {
    display: flex;
    flex: 1;
    padding: .25rem .5rem;
  }

  &__icon {
    display: flex;
    padding-right: .2rem;
    padding-top: .27rem;
    align-items: flex-start;
    justify-content: center;
  }

  &__content {
    @include typography-component-base;
    @include body2;

    display: flex;
    flex: 1;
    align-items: flex-start;
  }

  &__button-container {
    display: flex;
    align-items: flex-start;
    justify-content: center;
    padding-top: .27rem;

    &__button {
      display: flex;
      align-items: center;
      justify-content: center;
      height: 19px;
      width: 19px;
      padding: 0;
      border-radius: .2rem;
      cursor: pointer;
    }
  }

  &__actions-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 0 1rem;

    &__button {
      padding: .2rem;
    }
  }

  &--intent-primary {
    background-color: var(--lens-ui-intents-primary-dark);
    color: var(--lens-ui-intents-primary-dark-text);
    border: 1px solid var(--lens-ui-intents-primary-dark-border);

    [data-lens-element='toast__content'] {
      color: var(--lens-ui-intents-primary-dark-text);
    }

    svg,
    [data-lens-element='toast__icon'],
    [data-lens-element='toast__close-button'] {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-primary-dark-text);
      fill: var(--lens-ui-intents-primary-dark-text);
    }

    [data-lens-element='progress-bar'] {
      background-color: var(--lens-ui-intents-primary-darken);
    }

    [data-lens-element='progress-bar__indicator'] {
      background-color: #fff;
    }
  }

  &--intent-secondary {
    background-color: var(--lens-ui-intents-secondary-dark);
    color: var(--lens-ui-intents-secondary-dark-text);
    border: 1px solid var(--lens-ui-intents-secondary-dark-border);

    [data-lens-element='toast__content'] {
      color: var(--lens-ui-intents-secondary-dark-text);
    }

    svg,
    [data-lens-element='toast__icon'],
    [data-lens-element='toast__close-button'] {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-secondary-dark-text);
      fill: var(--lens-ui-intents-secondary-dark-text);
    }

    [data-lens-element='progress-bar'] {
      background-color: var(--lens-ui-intents-secondary-darken);
    }

    [data-lens-element='progress-bar__indicator'] {
      background-color: #fff;
    }
  }

  &--intent-success {
    background-color: var(--lens-ui-intents-success-dark);
    color: var(--lens-ui-intents-success-dark-text);
    border: 1px solid var(--lens-ui-intents-success-dark-border);

    [data-lens-element='toast__content'] {
      color: var(--lens-ui-intents-success-dark-text);
    }

    svg,
    [data-lens-element='toast__icon'],
    [data-lens-element='toast__close-button'] {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-success-dark-text);
      fill: var(--lens-ui-intents-success-dark-text);
    }

    [data-lens-element='progress-bar'] {
      background-color: var(--lens-ui-intents-success-darken);
    }

    [data-lens-element='progress-bar__indicator'] {
      background-color: #fff;
    }
  }

  &--intent-info {
    background-color: var(--lens-ui-intents-info-dark);
    color: var(--lens-ui-intents-info-dark-text);
    border: 1px solid var(--lens-ui-intents-info-dark-border);

    [data-lens-element='toast__content'] {
      color: var(--lens-ui-intents-info-dark-text);
    }

    svg,
    [data-lens-element='toast__icon'],
    [data-lens-element='toast__close-button'] {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-info-dark-text);
      fill: var(--lens-ui-intents-info-dark-text);
    }

    [data-lens-element='progress-bar'] {
      background-color: var(--lens-ui-intents-info-darken);
    }

    [data-lens-element='progress-bar__indicator'] {
      background-color: #fff;
    }
  }

  &--intent-warning {
    background-color: var(--lens-ui-intents-warning-dark);
    color: var(--lens-ui-intents-warning-dark-text);
    border: 1px solid var(--lens-ui-intents-warning-dark-border);

    [data-lens-element='toast__content'] {
      color: var(--lens-ui-intents-warning-dark-text);
    }

    svg,
    [data-lens-element='toast__icon'],
    [data-lens-element='toast__close-button'] {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-warning-dark-text);
      fill: var(--lens-ui-intents-warning-dark-text);
    }

    [data-lens-element='progress-bar'] {
      background-color: var(--lens-ui-intents-warning-darken);
    }

    [data-lens-element='progress-bar__indicator'] {
      background-color: #fff;
    }
  }

  &--intent-danger {
    background-color: var(--lens-ui-intents-danger-dark);
    color: var(--lens-ui-intents-danger-dark-text);
    border: 1px solid var(--lens-ui-intents-danger-dark-border);

    [data-lens-element='toast__content'] {
      color: var(--lens-ui-intents-danger-dark-text);
    }

    svg,
    [data-lens-element='toast__icon'],
    [data-lens-element='toast__close-button'] {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-danger-dark-text);
      fill: var(--lens-ui-intents-danger-dark-text);
    }

    [data-lens-element='progress-bar'] {
      background-color: var(--lens-ui-intents-danger-darken);
    }

    [data-lens-element='progress-bar__indicator'] {
      background-color: #fff;
    }
  }
}
