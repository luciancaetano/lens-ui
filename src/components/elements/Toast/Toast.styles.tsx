import styled from 'styled-components';

export const ToastContainer = styled.div`
  pointer-events: all;
  min-width: 300px;
  border-radius: .1rem;
  margin-top: .3rem;
  box-shadow: 0px 0px 3px #0000004d;
  display: flex;
  flex-direction: column;
  overflow: hidden;

  .lens-ui-progress-bar,
  .lens-ui-progress-bar-progress {
    border-radius: 0;
    height: .3rem;
  }

  .lens-ui-toasts-toast-main {
    display: flex;
    flex: 1;
    padding: .25rem .5rem;
  }

  .lens-ui-toasts-toast-icon {
    display: flex;
    padding-right: .2rem;
    align-items: center;
    justify-content: center;
  }

  .lens-ui-toasts-toast-content {
    display: flex;
    padding: .2rem;
    flex: 1;
    font-size: .8rem;
    align-items: center;
  }

  &.mobile {
    width: 100%;
  }

  .toast-close-button-container {
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;

    .toast-close-button {
      display: flex;
      align-items: center;
      justify-content: center;
      height: 19px;
      width: 19px;
      padding: 0;
      border-radius: .2rem;
    }
  }

  .toast-close-actions-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 0 1rem;

    .lens-ui-button {
      padding: .2rem .2rem;
    }
  }

  &.intent-primary {
    background-color: var(--lens-ui-intents-primary-dark);
    color: var(--lens-ui-intents-primary-dark-text);
    border: 1px solid var(--lens-ui-intents-primary-dark-border);

    .toast-close-button {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-primary-dark-text);
    }

    .lens-ui-progress-bar {
      background-color: var(--lens-ui-intents-primary-darken);
    }

    .lens-ui-progress-bar-progress {
      background-color: #fff;
    }
  }

  &.intent-secondary {
    background-color: var(--lens-ui-intents-secondary-dark);
    color: var(--lens-ui-intents-secondary-dark-text);
    border: 1px solid var(--lens-ui-intents-secondary-dark-border);

    .toast-close-button {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-secondary-dark-text);
    }

    .lens-ui-progress-bar {
      background-color: var(--lens-ui-intents-secondary-darken);
    }

    .lens-ui-progress-bar-progress {
      background-color: #fff;
    }
  }

  &.intent-success {
    background-color: var(--lens-ui-intents-success-dark);
    color: var(--lens-ui-intents-success-dark-text);
    border: 1px solid var(--lens-ui-intents-success-dark-border);

    .toast-close-button {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-success-dark-text);
    }

    .lens-ui-progress-bar {
      background-color: var(--lens-ui-intents-success-darken);
    }

    .lens-ui-progress-bar-progress {
      background-color: #fff;
    }
  }

  &.intent-info {
    background-color: var(--lens-ui-intents-info-dark);
    color: var(--lens-ui-intents-info-dark-text);
    border: 1px solid var(--lens-ui-intents-info-dark-border);

    .toast-close-button {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-info-dark-text);
    }

    .lens-ui-progress-bar {
      background-color: var(--lens-ui-intents-info-darken);
    }

    .lens-ui-progress-bar-progress {
      background-color: #fff;
    }
  }

  &.intent-warning {
    background-color: var(--lens-ui-intents-warning-dark);
    color: var(--lens-ui-intents-warning-dark-text);
    border: 1px solid var(--lens-ui-intents-warning-dark-border);

    .toast-close-button {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-warning-dark-text);
    }

    .lens-ui-progress-bar {
      background-color: var(--lens-ui-intents-warning-darken);
    }

    .lens-ui-progress-bar-progress {
      background-color: #fff;
    }
  }

  &.intent-danger {
    background-color: var(--lens-ui-intents-danger-dark);
    color: var(--lens-ui-intents-danger-dark-text);
    border: 1px solid var(--lens-ui-intents-danger-dark-border);

    .toast-close-button {
      background-color: transparent;
      border: 0;
      color: var(--lens-ui-intents-danger-dark-text);
    }

    .lens-ui-progress-bar {
      background-color: var(--lens-ui-intents-danger-darken);
    }

    .lens-ui-progress-bar-progress {
      background-color: #fff;
    }
  }
`;
