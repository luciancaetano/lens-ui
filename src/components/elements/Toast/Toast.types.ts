import React from 'react';
import { ITestableProps } from '../../../types';
import { IToastData } from '../../providers/ToastProvider/ToastProvider.types';

export interface IToastProps extends ITestableProps, React.HtmlHTMLAttributes<HTMLElement> {
  data: IToastData;
}
