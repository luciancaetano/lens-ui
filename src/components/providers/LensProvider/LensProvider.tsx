import React from 'react';
import DeviceProvider from '../DeviceProvider/DeviceProvider';
import AlertProvider from '../AlertProvider/AlertProvider';
import ModalProvider from '../ModalProvider/ModalProvider';
import ToastProvider from '../ToastProvider/ToastProvider';
import LocaleProvider from '../LocaleProvider/LocaleProvider';
import { ILensProviderProps } from './LensProvider.types';
import { DefaultLocales } from '../../../i18n/index';
import { CLASSES } from '../../..';

const LensProvider = ({
  children, deviceContextUpdateDebounceTime = 100,
  initialLocale = DefaultLocales.enUs,
  alertSettings = {
    responseLimit: 10,
  },
  toastSettings = {
    placement: 'bottom-right',
  },
}: React.PropsWithChildren<ILensProviderProps>) => (
  <DeviceProvider debounceTime={deviceContextUpdateDebounceTime}>
    <LocaleProvider initialLocale={initialLocale}>
      <AlertProvider {...alertSettings}>
        <ToastProvider {...toastSettings}>
          <ModalProvider>
            <div id={CLASSES.PortalRootContainer} />
            {children}
          </ModalProvider>
        </ToastProvider>
      </AlertProvider>
    </LocaleProvider>
  </DeviceProvider>
);

export default LensProvider;
