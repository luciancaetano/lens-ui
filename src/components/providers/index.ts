export { default as DeviceProvider } from './DeviceProvider/DeviceProvider';
export { default as ToastProvider } from './ToastProvider/ToastProvider';

export * from './DeviceProvider/DeviceProvider.types';
export * from './ToastProvider/ToastProvider.types';
