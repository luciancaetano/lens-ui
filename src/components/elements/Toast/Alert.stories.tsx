/* eslint react/jsx-pascal-case: 0 */
import React from 'react';
import Button from './components/elements/Button/Button';
import LensProvider from './components/providers/LensProvider/LensProvider';
import { useAlert } from './hooks';

export default {
  title: '3. Hooks/Alert',
  excludeStories: /__.*$/,
  decorators: [
    (Story) => (<LensProvider><Story /></LensProvider>),
  ],
};

export const Alert = () => {
  const { addAlert, alertResults } = useAlert();

  return (
    <div>
      <div>
        o hook useAlert retorna um array [alertFn, results, data]<br />
        O Id de cada alerta é retornado na função alert, e suas repostas são guardas em um objeto para consulta<br />
        As respostas também podem ser consultadas pelo atributo promise do objeto retornado na função alert<br />
        <b>Note que estes alertas apenas armazenam respostas e nunca dados por segurança</b><br />
        <i>A prática do prompt neste design system não é recomendada, as questões de dados aos usuários devem ser feitas dentro de formulários</i>
        <br />
        <code>
          <pre>
            {JSON.stringify(alertResults, null, '\t')}
          </pre>
        </code>

      </div>
      <Button onClick={async () => {
        addAlert({
          text: 'Deseja mover para lixeira/excluir este registro?',
          settings: {
            showConfirmButton: true,
            showCancelButton: true,
            showDenyButton: true,
            cancelButtonText: 'Cancelar',
            confirmButtonText: 'Mover para lixeira',
            denyButtonText: 'Excluir',
          },
        });

        addAlert({
          text: 'Notificação simples',
        });
      }}
      >
        Show Alert
      </Button>
    </div>
  );
};
