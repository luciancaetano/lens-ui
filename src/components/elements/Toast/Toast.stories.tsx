/* eslint-disable react/destructuring-assignment */
import React from 'react';
import { ComponentMeta } from '@storybook/react';
import Toast from './Toast';
import LensProvider from '../../providers/LensProvider/LensProvider';
import '../../../styles';
import { useToast } from '../../../hooks';
import Button from '../Button/Button';
import Icon from '../Icon/Icon';

export default {
  title: '3. Hooks/useToast',
  component: Toast,
  decorators: [
    (Story) => <LensProvider><Story /></LensProvider>,
  ],
} as ComponentMeta<typeof Toast>;

export var _Toast = function () {
  const { addToast } = useToast();

  return (
    <Button
      intent="success"
      onClick={async () => {
        addToast({
          content: 'lorem impsum',
          icon: <Icon name="BsExclamationTriangleFill" />,
          intent: 'success',
          dismiss: 3000,
        });
      }}
    >
      Show Toast
    </Button>
  );
};
