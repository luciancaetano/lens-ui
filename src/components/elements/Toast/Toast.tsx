import clsx from 'clsx';
import isNumber from 'lodash/isNumber';
import clamp from 'lodash/clamp';
import map from 'lodash/map';
import React, {
  useRef, useEffect, useCallback, useMemo,
} from 'react';
import Icon from '../Icon/Icon';
import { useDevice, useToast } from '../../../hooks';
import styles from './Toast.module.scss';
import ProgressBar from '../ProgressBar/ProgressBar';
import { IToastProps } from './Toast.types';
import Button from '../Button/Button';

const INTERVAL = 250;

/**
 * Toast displays a short, important message in a way that attracts the user's attention without interrupting the user's task.
 */
const Toast:React.FC<IToastProps> = ({
  testingID, data,
}) => {
  const { removeToast } = useToast();
  const initialized = useRef(false);
  const interval = useRef<any>(INTERVAL);
  const timeProgress = useRef<number>(0);
  const progressBar = useRef<HTMLDivElement>(null);
  const { isPhone } = useDevice();

  const autoDismiss = useMemo(() => isNumber(data.dismiss), [data]);

  const updateProgress = useCallback((timePassed: number) => {
    if (!isNumber(data.dismiss) || !progressBar.current) return;

    const element = progressBar.current.querySelector<HTMLDivElement>('[data-lens-element=\'progress-bar__indicator\']');
    if (element) {
      element.style.width = `${clamp((timePassed / data.dismiss) * 100, 0, 100)}%`;
    }
  }, [data.dismiss]);

  useEffect(() => {
    if (!initialized.current) {
      initialized.current = true;
      if (isNumber(data.dismiss)) {
        interval.current = setInterval(() => {
          if (timeProgress.current >= data.dismiss) {
            clearInterval(interval.current);
            setTimeout(() => removeToast(data.id), INTERVAL);
          } else {
            timeProgress.current += INTERVAL;
            updateProgress(timeProgress.current);
          }
        }, INTERVAL);
      }
    }
  }, [data, removeToast, updateProgress]);

  const handleClose = useCallback(() => {
    removeToast(data.id);
  }, [removeToast, data]);

  const actions = useMemo(() => map(data.actions, ({ content, ...action }) => (
    <Button {...action} className={clsx(styles.toastActionsContainerButton, action.className)} size="small">{content}</Button>
  )), [data.actions]);

  return (
    <div
      data-lens-element="toast"
      data-lens-intent={data.intent}
      data-testid={testingID}
      className={clsx(styles.toast, isPhone && styles.toastMobile, styles[`toast--intent-${data.intent}`], data.className)}
    >
      <div className={styles.toastMain}>
        {data.icon && (<div className={styles.toastIcon}>{data.icon}</div>)}
        <div className={styles.toastContent} data-lens-element="toast__content">{data.content}</div>
        {actions.length > 0 && (
          <div data-lens-element="toast__actions" className={clsx(styles.toastActionsContainer, data.actionsContainerClassName)}>
            {actions}
          </div>
        )}
        <div className={styles.toastButtonContainer} onClick={handleClose}>
          <button data-lens-element="toast__close-button">
            <Icon name="BsXSquareFill" />
          </button>
        </div>
      </div>
      {autoDismiss && <ProgressBar progress={0} size="tiny" ref={progressBar} />}
    </div>
  );
};
export default Toast;
