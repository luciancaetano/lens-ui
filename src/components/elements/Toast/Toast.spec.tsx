import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import LensProvider from '../../providers/LensProvider/LensProvider';
import Toast from './Toast';
import { IToastData } from '../../providers';

const removeToast = jest.fn();

jest.mock('../../../hooks', () => ({
  useDevice: () => ({
    isDesktop: true,
    isPhone: false,
    isTablet: false,
    online: true,
    orientation: 'landscape',
    os: 'android',
    windowSize: {
      width: 1920,
      height: 1080,
    },
  }),
  useToast: () => ({
    removeToast,
  }),
}));

jest.useFakeTimers();

const data: IToastData = {
  content: 'toast_content',
  id: 'toast_id',
  intent: 'info',
  className: 'myCustomClass',
  icon: 'my_icon',
};

describe('<Toast/>', () => {
  beforeEach(() => {
    removeToast.mockClear();
  });

  it('render <Toast/>', async () => {
    const { container } = render(
      <LensProvider>
        <Toast data={data} />
      </LensProvider>,
    );

    expect(container.querySelector('.lens-ui-toasts-toast')).toMatchSnapshot();
  });

  it('render <Toast/> with invalid dismiss', async () => {
    const { container } = render(
      <LensProvider>
        <Toast data={{ ...data, dismiss: 'invalidvalue' as any }} />
      </LensProvider>,
    );

    expect(container.querySelector('.lens-ui-toasts-toast')).toMatchSnapshot();
  });

  it('render <Toast/> close toast', async () => {
    render(
      <LensProvider>
        <Toast data={{ ...data, dismiss: 300 }} />
      </LensProvider>,
    );
    jest.advanceTimersByTime(1000);

    fireEvent.click(window.document.querySelector('[data-lens-element="toast__close-button"]'));

    expect(removeToast).toHaveBeenCalled();
  });

  it('render <Toast/> autoDismiss toast', async () => {
    render(
      <LensProvider>
        <Toast data={{ ...data, dismiss: 300 }} />
      </LensProvider>,
    );
    jest.advanceTimersByTime(1000);

    expect(removeToast).toHaveBeenCalled();
  });
});
