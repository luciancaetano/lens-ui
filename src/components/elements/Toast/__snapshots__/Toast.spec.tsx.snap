// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toast/> render <Toast/> 1`] = `null`;

exports[`<Toast/> render <Toast/> with invalid dismiss 1`] = `null`;
